graphql_compiler/tests/test_compiler.py

```diff
@@ -2868,3 +2868,907 @@ def test_traverse_and_optional_and_traverse(self):
+                    .ifThenElse{it == null}{null}{it.out('Animal_OfSpecies')}
+                    .as('Animal__out_Animal_ParentOf__out_Animal_OfSpecies___1')
+                .back('Animal__out_Animal_ParentOf___1')
+            .optional('Animal___2')
+            .as('Animal___3')
+            .transform{it,
+                m -> new com.orientechnologies.orient.core.record.impl.ODocument([
+                animal_name: m.Animal___1.name,
+                child_name: (
+                    (m.Animal__out_Animal_ParentOf___1 != null) ?
+                        m.Animal__out_Animal_ParentOf___1.name : null
+                ),
+                child_species: (
+                    (m.Animal__out_Animal_ParentOf__out_Animal_OfSpecies___1 != null) ?
+                        m.Animal__out_Animal_ParentOf__out_Animal_OfSpecies___1.name : null
+                ),
+                parent_name: (
+                    (m.Animal__in_Animal_ParentOf___1 != null) ?
+                        m.Animal__in_Animal_ParentOf___1.name : null
+                ),
+                sibling_name: (
+                    (m.Animal__in_Animal_ParentOf__out_Animal_ParentOf___1 != null) ?
+                        m.Animal__in_Animal_ParentOf__out_Animal_ParentOf___1.name : null
+                )
+            ])}
+        '''
+
+        check_test_data(self, test_data, expected_match, expected_gremlin)
+
+    def test_optional_traversal_and_optional_without_traversal(self):
+        test_data = test_input_data.optional_traversal_and_optional_without_traversal()
+
+        expected_match = '''
+            SELECT EXPAND($result)
+            LET
+            $optional__0 = (
+                SELECT Animal___1.name AS `animal_name`,
+                        if(
+                            eval("(Animal__in_Animal_ParentOf___1 IS NOT null)"),
+                            Animal__in_Animal_ParentOf___1.name,
+                            null
+                        ) AS `parent_name`
+                    FROM (
+                        MATCH {{
+                            class: Animal,
+                            where: ((
+                                (
+                                    (out_Animal_ParentOf IS null)
+                                    OR
+                                    (out_Animal_ParentOf.size() = 0)
+                                )
+                                AND
+                                (name LIKE ('%' + ({wanted} + '%')))
+                                )),
+                            as: Animal___1
+                        }}.in('Animal_ParentOf') {{
+                            optional: true,
+                            as: Animal__in_Animal_ParentOf___1
+                        }} ,
+                        {{
+                            class: Animal,
+                            as: Animal___1
+                        }}
+                        RETURN $matches
+                    )
```

Comment:
Note how here there's a double de-indent -- that means that the indentation is messed up somewhere, since all of the paren + de-indents should come in one at a time. This issue appears several more times in this file.